### PR TITLE
[v6r20] MJF utility and Watchdog StopSig mechanism

### PR DESCRIPTION
**Pull request by Andrew-McNab-UK**
BEGINRELEASENOTES
*Core
NEW: MJF utility added, providing a general interface to Machine/Job Features values.

*WorkloadManagementSystem
CHANGE: dirac-wms-cpu-normalization uses the abstracted DB12 benchmark script used by the HEPiX Benchmarking Working Group, and the new MJF utility to obtain values from the system and to save them into the DIRAC LocalSite configuration.
NEW: StopSigRegex, StopSigStartSeconds, StopSigFinishSeconds, StopSigNumber added to JDL, which cause Watchdog to send a signal StopSigNumber to payload processes matching StopSigRegex when there are less than StopSigFinishSeconds of wall clock remaining according to MJF.
ENDRELEASENOTES

These changes to the Watchdog will be followed by a PR with separate changes to make the JobAgent and TimeLeft module use the new MJF utility too.

### CODE REVIEW

File: WorkloadManagementSystem/JobWrapper/Watchdog.py [hunk 1]
```diff
@@ -20,22 +20,45 @@
...
+from DIRAC.Core.Utilities.Subprocess                    import getChildrenPIDs
 
 class Watchdog( object ):
 
   #############################################################################
-  def __init__( self, pid, exeThread, spObject, jobCPUtime, memoryLimit = 0, processors = 1, systemFlag = 'linux' ):
+  def __init__( self, pid, exeThread, spObject, jobCPUtime, jobArgs, memoryLimit = 0, processors = 1, systemFlag = 'linux' ):
     """ Constructor, takes system flag as argument.
     """
+    if "StopSigStartSeconds" in jobArgs:
```
Comment (andresailer) [inline WorkloadManagementSystem/JobWrapper/Watchdog.py]: All these checks can be, e.g., 
```
self.stopSigStartSeconds = int( jobArgs.get('StopSigStartSeconds', 30*60) )
```

File: WorkloadManagementSystem/scripts/dirac-wms-cpu-normalization.py [hunk 2]
```diff
@@ -32,43 +32,48 @@
...
+  mjf = MJF.MJF()
+  mjf.updateConfig()
+  
+  db12JobFeature = mjf.getJobFeature( 'db12' )
+  hs06JobFeature = mjf.getJobFeature( 'hs06' )
+
+  result = singleDiracBenchmark( 1 )
+
+  if result is None:
+    gLogger.error( 'Cannot make benchmark measurements' )
```
Comment (andresailer) [inline WorkloadManagementSystem/scripts/dirac-wms-cpu-normalization.py]: Should exit `if result is None` or there will be an exception in line 51

File: WorkloadManagementSystem/JobWrapper/Watchdog.py [hunk 3]
```diff
@@ -20,22 +20,45 @@
...
 from DIRAC.ConfigurationSystem.Client.PathFinder        import getSystemInstance
 from DIRAC.Core.Utilities.ProcessMonitor                import ProcessMonitor
 from DIRAC.Core.Utilities.TimeLeft.TimeLeft             import TimeLeft
+from DIRAC.Core.Utilities.Subprocess                    import getChildrenPIDs
 
 class Watchdog( object ):
 
   #############################################################################
-  def __init__( self, pid, exeThread, spObject, jobCPUtime, memoryLimit = 0, processors = 1, systemFlag = 'linux' ):
+  def __init__( self, pid, exeThread, spObject, jobCPUtime, jobArgs, memoryLimit = 0, processors = 1, systemFlag = 'linux' ):
```
Comment (fstagni) [inline WorkloadManagementSystem/JobWrapper/Watchdog.py]: This is the only thing that makes me a bit scared. I would prefer that the change of signature would be:

```
def __init__( self, pid, exeThread, spObject, jobCPUtime, memoryLimit = 0, processors = 1, systemFlag = 'linux', jobArgs = {} ):
```

File: WorkloadManagementSystem/JobWrapper/Watchdog.py [hunk 4]
```diff
@@ -20,22 +20,45 @@
 __RCSID__ = "$Id$"
 
 import os
+import re
 import time
 
 from DIRAC                                              import S_OK, S_ERROR, gLogger
 from DIRAC.Core.Utilities                               import Time
+from DIRAC.Core.Utilities                               import MJF
```
Comment (fstagni) [inline WorkloadManagementSystem/JobWrapper/Watchdog.py]: This is maybe a matter of taste, but in 99% of the cases we do:
```
from DIRAC.Core.Utilities.MJF                               import MJF
```

**Comment by fstagni** [comment]
@phicharp would you mind adding your review? Thanks

File: WorkloadManagementSystem/Client/DIRACbenchmark.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+           'geometric_mean'  : product ** (1.0 / copies),
+           'median'          : raw[(copies-1) / 2] }
+  
+def wholenodeDiracBenchmark( copies = None, iterations = 1, extraIteration = False ): 
+
+  """ Run as many copies as needed to occupy the whole machine
+  """
+  
+  # Try $MACHINEFEATURES first if not given by caller
+  if not copies and 'MACHINEFEATURES' in os.environ:
```
Comment (phicharp) [inline WorkloadManagementSystem/Client/DIRACbenchmark.py]: `if copies is None` ?

File: WorkloadManagementSystem/JobWrapper/Watchdog.py (hunk 1 above)
Comment (phicharp) [reply WorkloadManagementSystem/JobWrapper/Watchdog.py]: Agree, much cleaner

**Comment by phicharp** [comment]
Sounds OK to me, but style is definitely not pep8 ;-)
I think for new code or modified code we should try and stick to it, except the tab length (2) & line length (12)

**Comment by fstagni** [comment]
Now pylint fails because:

************* Module DIRAC.WorkloadManagementSystem.JobWrapper.Watchdog
WorkloadManagementSystem/JobWrapper/Watchdog.py:224: [E1101(no-member), Watchdog._performWallClockChecks] Class 'MJF' has no 'MJF' member
************* Module dirac-wms-cpu-normalization
WorkloadManagementSystem/scripts/dirac-wms-cpu-normalization.py:40: [E1101(no-member), ] Class 'MJF' has no 'MJF' member

**Comment by fstagni** [comment]
IMHO review OK, if also @phicharp agrees.

File: WorkloadManagementSystem/JobWrapper/WatchdogLinux.py [hunk 6]
```diff
@@ -22,10 +22,10 @@
 
 class WatchdogLinux( Watchdog ):
 
-  def __init__( self, pid, thread, spObject, jobCPUtime, memoryLimit = 0, processors = 1, systemFlag = 'linux' ):
+  def __init__( self, pid, thread, spObject, jobCPUtime, memoryLimit = 0, processors = 1, systemFlag = 'linux', jobArgs = {} ):
     """ Constructor, takes system flag as argument.
     """
-    Watchdog.__init__( self, pid, thread, spObject, jobCPUtime, memoryLimit, processors, systemFlag )
+    Watchdog.__init__( self, pid, thread, spObject, jobCPUtime, memoryLimit, processors, systemFlag, jobArgs )
```
Comment (phicharp) [inline WorkloadManagementSystem/JobWrapper/WatchdogLinux.py]: Use the names of arguments and not position?

File: WorkloadManagementSystem/JobWrapper/WatchdogMac.py [hunk 7]
```diff
@@ -21,10 +21,10 @@
 
 class WatchdogMac( Watchdog ):
 
-  def __init__( self, pid, thread, spObject, jobCPUtime, memoryLimit = 0, processors = 1, systemFlag = 'mac' ):
+  def __init__( self, pid, thread, spObject, jobCPUtime, memoryLimit = 0, processors = 1, systemFlag = 'mac', jobArgs = {} ):
     """ Constructor, takes system flag as argument.
     """
-    Watchdog.__init__( self, pid, thread, spObject, jobCPUtime, memoryLimit, processors, systemFlag )
+    Watchdog.__init__( self, pid, thread, spObject, jobCPUtime, memoryLimit, processors, systemFlag, jobArgs )
```
Comment (phicharp) [inline WorkloadManagementSystem/JobWrapper/WatchdogMac.py]: id

File: WorkloadManagementSystem/JobWrapper/test/Test_JobWrapper.py [hunk 8]
```diff
@@ -66,7 +66,7 @@ def test_InputData( self ):
     self.assertTrue( res['OK'] )
 
   def test__performChecks( self ):
-    wd = WatchdogLinux( os.getpid(), MagicMock(), MagicMock(), 1000, 1024 * 1024 )
+    wd = WatchdogLinux( os.getpid(), MagicMock(), MagicMock(), 1000, 1024 * 1024, jobArgs = { 'StopSigNumber' : 10 } )
```
Comment (phicharp) [inline WorkloadManagementSystem/JobWrapper/test/Test_JobWrapper.py]: as  well..

Comment (Andrew-McNab-UK) [reply WorkloadManagementSystem/JobWrapper/test/Test_JobWrapper.py]: Thanks: I've made all the Watchdog*** calls use named arguments now and made the argument names consistent in the signatures (thread vs exeThread)

**Pull request merged**